### PR TITLE
Constrain memory ordering on Bank::freeze_started

### PR DESCRIPTION
**Pull request by brooksprumo**
#### Problem

Please refer to PR #21311 for much more background on atomics and memory ordering.

`Bank::freeze_started` is used for logic/control flow (to make sure the bank's internal state has been hash, and to mark the slot as frozen in AccountsDb), but is not constrained for this (to prevent reordering/guarantee ordering).

#### Summary of Changes

* Constrain with `Release` and `Acquire`.
* Rename the fn that checks if freezing has started to include `is_` in the name.

### CODE REVIEW

**Review by carllin — APPROVED** [review]

**Comment by sakridge** [comment]
What other atomic access does it need to be not ordered around?

**Comment by brooksprumo** [comment]
> What other atomic access does it need to be not ordered around?

This is a good question. Looks like most of the callers of `is_freeze_started()` just assert its _not_ freezing before going on. For example, `commit_transactions()` check `is_freeze_started()` before atomically incrementing transaction count and signature count. So doesn't look like these other atomics matter.

So then I'm inferring that `is_freeze_started()` is meant to be a performance optimization over `is_frozen()`, so (1) the RwLock doesn't need to be grabbed, and (2) don't have to wait for `hash_internal_state()` to run first (even though that's not guaranteed by the compiler/cpu) before checking freeze state. Is that right, @sakridge ? That and we need something that supports interior mutability due to Bank being sent to other threads.

Looking at the metrics, `hash_internal_state()`'s scan hash time has an average of < 100 ms, which is non-negligible. But it looks like all the places where `is_freeze_started()` is called are just checking invariants via `assert!()`.

To summarize, I think `Relaxed` here is fine. But also I think `freeze_started` can be removed, and callers of `is_freeze_started()` can use `is_frozen()` instead.

**Comment by brooksprumo** [comment]
Closing this PR, as I do not intend to merge it.
